numpy/core/tests/test_datetime.py

```diff
@@ -740,19 +758,58 @@ def test_pyobject_roundtrip(self):
         for unit in ['M8[as]', 'M8[16fs]', 'M8[ps]', 'M8[us]',
                      'M8[300as]', 'M8[20us]']:
             b = a.copy().view(dtype=unit)
-            b[0] = '-0001-01-01T00'
-            b[1] = '-0001-12-31T00'
-            b[2] = '0000-01-01T00'
-            b[3] = '0001-01-01T00'
+            # 1e-7 seconds from epoch:
+            b[0] = '1969-12-31T23:59:59.999999'
+            # 9.0 seconds from epoch, near as unit span limit:
+            b[1] = '1969-12-31T23:59:51.0'
+            # similar thing on other side of epoch:
+            b[2] = '1970-01-01T00:00:09.19'
+            # could perhaps use more creative values
+            # here, but must be cautious to respect
+            # allowable spans following fix to gh-5452
+            b[3] = '1969-12-31T23:59:59.999999'
             b[4] = '1969-12-31T23:59:59.999999'
             b[5] = '1970-01-01T00'
-            b[6] = '9999-12-31T23:59:59.999999'
-            b[7] = '10000-01-01T00'
+            b[6] = '1970-01-01T00'
+            b[7] = '1970-01-01T00'
             b[8] = 'NaT'
 
             assert_equal(b.astype(object).astype(unit), b,
                             "Error roundtripping unit %s" % unit)
 
+    @pytest.mark.parametrize("time_str", [
+        # original example from gh-5452
+        # sub-ns units / precision can't handle multiple year
+        # time spans from the epoch
+        ('2015-01-01T01:00:00.0000000000'),
+        # test a value just beyond the ps threshold of 106 days 
+        # we currently reject on the seconds resolution for ps
+        ('1970-04-17T00:00:01.0000000000'),
+        # likewise for fs threshold of 2.6 hours
+        # we currently reject on the seconds resolution for fs
+        ('1970-01-01T02:36:01.000000000000000'),
+        # likewise for as threshold of 9.2 seconds, also
+        # resolved on seconds resolution
+        # NOTE: may want to increase resolution here a bit in
+        # future
+        ('1970-01-01T00:00:10.000000000000000000'),
+        ])
+    def test_span_limit_handling(self, time_str):
+        # regression test for gh-5452
+        with assert_raises_regex(ValueError, 'Time span'):
+            numpy.datetime64(time_str)
+
+    @pytest.mark.xfail(reason="work in progress")
+    @pytest.mark.parametrize("time_unit, time_val", [
+        ('ns', np.datetime64(int(np.iinfo(np.int64).max * (10 ** -9)), 's') + np.timedelta64(0, 'ns')),
+        ('ps', np.datetime64(int(np.iinfo(np.int64).max * (10 ** -12)), 's') + np.timedelta64(0, 'ps')),
```

Comment:
This seems to detect the precision reduction Eric is worried about in this PR (this test now correctly fails), but I'm slightly concerned about the `int()` call here -- it truncates the float I get from the division to produce the number of seconds from the epoch.

Obviously, I can't give a float to intialize the `datetime64`. Is there a better approach? Python `decimal` module or something to flush through the strictest precision test possible, or is this going to be close enough?

Reply:
Use `max // (10**12)` instead, perhaps?

Reply:
Or perhaps `(max + 10**12 - 1) // (10**12)`, which rounds up not down

Reply:
My best attempts so far use `divmod` to recuperate the remainders and add them back in using appropriate units.

However, I've grown the unit test quite a bit now, and expanding to include symmetric time spans seems to indicate silent wrap-around for a subset of the time units:

```python
input_seconds, remainder_ps = divmod(np.iinfo(np.int64).max, 10**12)
input_str_minus = str(np.datetime64(0, 's') - np.timedelta64(input_seconds, 's') - np.timedelta64(remainder_ps, 'ps'))
input_str_plus = str(np.datetime64(0, 's') + np.timedelta64(input_seconds, 's') + np.timedelta64(remainder_ps, 'ps'))
print("minus:", np.datetime64(input_str_minus))
print("plus:", np.datetime64(input_str_plus))
```

Both end up after the epoch, even on master branch:
```
minus: 1970-04-16T05:57:07.963145224193 # looks wrong
plus: 1970-04-17T18:02:52.036854775807 # looks about right
```

Is the asymmetry surprising? Our docs would suggest there should be no asymmetry. I originally noticed that pushing just slightly beyond the time spans in the positive directions tends to produce `NaT`, so I wanted to have a secondary check in the unit test to confirm that I was actually very close to the "precision" limit.